Modules/_elementtree.c

```diff
@@ -1632,46 +1632,59 @@ static PyObject *
 _elementtree_Element_remove_impl(ElementObject *self, PyObject *subelement)
 /*[clinic end generated code: output=38fe6c07d6d87d1f input=6133e1d05597d5ee]*/
 {
-    Py_ssize_t i;
-    int rc;
-    PyObject *found;
-
-    if (!self->extra) {
+    if (self->extra == NULL) {
         /* element has no children, so raise exception */
-        PyErr_SetString(
-            PyExc_ValueError,
-            "list.remove(x): x not in list"
-            );
-        return NULL;
+        goto not_found;
     }
 
-    for (i = 0; i < self->extra->length; i++) {
-        if (self->extra->children[i] == subelement)
+    Py_ssize_t i;
+    const Py_ssize_t length = self->extra->length;
+    for (i = 0; self->extra && i < self->extra->length; i++) {
+        if (self->extra->children[i] == subelement) {
             break;
-        rc = PyObject_RichCompareBool(self->extra->children[i], subelement, Py_EQ);
-        if (rc > 0)
+        }
+        PyObject *child = Py_NewRef(self->extra->children[i]);
+        int rc = PyObject_RichCompareBool(child, subelement, Py_EQ);
+        Py_DECREF(child);
+        if (self->extra ==  NULL || self->extra->length != length) {
+            goto fail;
+        }
+        if (rc > 0) {
             break;
-        if (rc < 0)
+        }
+        else if (rc < 0) {
             return NULL;
+        }
     }
 
-    if (i >= self->extra->length) {
+    // An extra check must be done if the mutation occurs at the very last
+    // step and removes or clears the 'extra' list (the condition on the
+    // length would not be satisfied any more).
+    if (self->extra == NULL || self->extra->length != length) {
+        goto fail;
+    }
+    else if (i >= self->extra->length) {
         /* subelement is not in children, so raise exception */
-        PyErr_SetString(
-            PyExc_ValueError,
-            "list.remove(x): x not in list"
-            );
-        return NULL;
+        goto not_found;
     }
 
-    found = self->extra->children[i];
+    PyObject *found = self->extra->children[i];
 
     self->extra->length--;
-    for (; i < self->extra->length; i++)
+    for (; i < self->extra->length; i++) {
         self->extra->children[i] = self->extra->children[i+1];
+    }
 
     Py_DECREF(found);
     Py_RETURN_NONE;
+
+not_found:
+    PyErr_SetString(PyExc_ValueError, "list.remove(x): x not in list");
+    return NULL;
+
+fail:
+    PyErr_SetString(PyExc_RuntimeError, "children mutated during iteration");
```

Comment:
I do not think that we need to distinguish this case from the "not found" case. We can't guarantee that this exception is always raised if children mutated during iteration -- it is not raised if a subelement was added and removed during iteration. And the Python implementation does not detect this. So it is better to not complicate the code without the need.

Reply:
What if there is a side effect where the object is marked as not being in the list (e.g., by swapping an element that has been checked with the element to remove)? do we care that the message would tell me "not in the list" and that after I print the node, I would see it because of the modification? (actually, I think that's the reason why I had the versioning).

My point was to raise a RuntimeError instead of a ValueError to indicate that sometimes bad at runtime happened. And I think that's what dicts do (but I haven't checked if lists also raise a RuntimeError in this case).

Reply:
No, we should not care. The list can be modified after raising an exception.

We cannot detect all cases when "something happened". Python implementation does not detect this at all. Let's do not complicate the code without need.

Dict is a different story, because modifying a dict is more unpredicable than modifying a list.

Reply:
Ok, then I'm ok with just re-using the same error path. It would also simplify the tests by the way (and I think it's an example of "doing weird stuff -> getting weird stuff"). I'll just write something that avoids a crash (but tomorrow).